packages/git/src/browser/git-decoration-provider.ts

```diff
@@ -14,54 +14,103 @@
-            this.onGitEvent(event);
-        });
+    protected async handleGitEvent(event: GitStatusChangeEvent | undefined): Promise<void> {
+        this.changeEvent = event;
+        this.updateDecorations();
     }
 
-    private async onGitEvent(event: GitStatusChangeEvent | undefined): Promise<void> {
-        if (!event) {
+    protected updateDecorations(): void {
+        if (!this.changeEvent) {
             return;
         }
-
         const newDecorations = new Map<string, Decoration>();
-        this.collectDecorationData(event.status.changes, newDecorations);
+        this.collectDecorationData(this.changeEvent.status.changes, newDecorations);
 
-        const uris = new Set([...this.decorations.keys()].concat([...newDecorations.keys()]));
+        this.uris = new Set([...this.decorations.keys()].concat([...newDecorations.keys()]));
         this.decorations = newDecorations;
-        this.onDidChangeDecorationsEmitter.fire(Array.from(uris, value => new URI(value)));
+        this.triggerDecorationChange();
     }
 
-    private collectDecorationData(changes: GitFileChange[], bucket: Map<string, Decoration>): void {
+    protected collectDecorationData(changes: GitFileChange[], bucket: Map<string, Decoration>): void {
         changes.forEach(change => {
             const color = GitFileStatus.getColor(change.status, change.staged);
             bucket.set(change.uri, {
                 bubble: true,
-                colorId: color.substring(12, color.length - 1).replace(/-/g, '.'),
+                colorId: this.colorsEnabled ? color.substring(12, color.length - 1).replace(/-/g, '.') : undefined,
                 tooltip: GitFileStatus.toString(change.status),
                 letter: GitFileStatus.toAbbreviation(change.status, change.staged)
             });
         });
     }
 
     provideDecorations(uri: URI, token: CancellationToken): Decoration | Promise<Decoration | undefined> | undefined {
-        return this.decorations.get(uri.toString());
+        if (this.decorationsEnabled) {
+            return this.decorations.get(uri.toString());
+        }
+    }
+
+    protected handlePreferenceChange(event: PreferenceChangeEvent<GitConfiguration>): void {
+        const { preferenceName, newValue } = event;
+        if (preferenceName === 'git.decorations.enabled' || preferenceName === 'git.decorations.colors') {
+            if (preferenceName === 'git.decorations.enabled') {
+                const decorationsEnabled = !!newValue;
+                if (this.decorationsEnabled !== decorationsEnabled) {
+                    this.decorationsEnabled = decorationsEnabled;
+                }
+            }
+            if (preferenceName === 'git.decorations.colors') {
+                const colorsEnabled = !!newValue;
+                if (this.colorsEnabled !== colorsEnabled) {
+                    this.colorsEnabled = colorsEnabled;
+                }
+            }
+            this.updateDecorations();
+        }
```

Comment:
Suggestion: Since all these comparisons can quickly become difficult to maintain we should maybe opt for something like this. It effectively cuts the amount of string comparisons in half.
```suggestion
        let updateDecorations = false;
        if (preferenceName === 'git.decorations.enabled') {
            updateDecorations = true;
            const decorationsEnabled = !!newValue;
            if (this.decorationsEnabled !== decorationsEnabled) {
                this.decorationsEnabled = decorationsEnabled;
            }
        }
        if (preferenceName === 'git.decorations.colors') {
            updateDecorations = true;
            const colorsEnabled = !!newValue;
            if (this.colorsEnabled !== colorsEnabled) {
                this.colorsEnabled = colorsEnabled;
            }
        }
        if (updateDecorations) {
            this.updateDecorations();
        }
```